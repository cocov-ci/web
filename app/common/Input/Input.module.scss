@import "styles/variables";

.base {
  position: relative;
  display: inline-flex;
  align-items: center;
  width: 100%;
}

.label {
  margin-right: 8px;

  &.labelDisabled {
    opacity: 0.4;
  }
}

.input {
  background: $colorInputBackground;
  border: 1px solid $colorInputBorder;
  border-radius: 4px;
  font-size: $fontInput;
  padding: 5px 10px 4px;
  box-sizing: border-box;
  width: 100%;

  &::placeholder {
    color: $colorInputPlaceholder;
  }

  &:focus {
    outline: 0;
    border: 1px solid $colorBlue;
    background: $colorInputFocusBackground;
    box-shadow: 0 0 5px 1px $colorInputFocusShadow;
  }

  &.errored {
    border: 1px solid $colorInputErrorBorder;

    &:focus {
      box-shadow: 0 0 5px 1px $colorInputErrorFocusShadow;
    }
  }

  &.withIcon {
    padding: 8px 0 8px 30px;
    font-size: 17px;
    border-radius: 8px;
  }

  &.withSpinner {
    padding-right: 30px;
  }
}

.iconContainer {
  line-height: 15px;
  position: relative;
  margin-right: -18px;
  left: 8px;
  top: 1px;
}

.loading {
  position: absolute;
  top: 50%;
  transform: translateY(-20px);
  right: 8px;
}
