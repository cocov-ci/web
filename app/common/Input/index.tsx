'use client'

import classNames from 'classnames'
import { LucideIcon } from 'lucide-react'
import React, { ChangeEventHandler, useRef, useState } from 'react'

import Loading from '../Loading'

import styles from './Input.module.scss'

interface InputProps {
  icon?: LucideIcon
  className?: string
  value?: string
  disabled?: boolean
  placeholder?: string
  onChange?: ChangeEventHandler<HTMLInputElement>
  errored?: boolean
  type: React.HTMLInputTypeAttribute
  label?: string
  width?: string
  loading?: boolean
  variation?: 'light' | 'dark'
  onKeyDown?: (event: React.KeyboardEvent<HTMLInputElement>) => void
  onKeyUp?: (event: React.KeyboardEvent<HTMLInputElement>) => void
  autoFocus?: boolean
  innerRef?: React.MutableRefObject<HTMLInputElement | null>
}

const Input = ({
  className,
  value,
  disabled,
  placeholder,
  onChange,
  errored,
  type,
  label,
  loading,
  width = '100%',
  variation = 'light',
  onKeyUp,
  onKeyDown,
  autoFocus,
  icon: Icon,
  innerRef,
}: InputProps) => {
  const inputRef = useRef<HTMLInputElement | null>(null)
  const [focused, setFocused] = useState(false)

  const updateFocusState = () => {
    if (inputRef.current && document.activeElement !== null) {
      setFocused(document.activeElement === inputRef.current)
    }
  }

  const captureFocus = () => inputRef.current && inputRef.current.focus()

  return (
    <div
      className={classNames(styles.base, {
        [styles.light]: variation == 'light',
        [styles.dark]: variation == 'dark',
      })}
      style={{ width }}
    >
      {label && (
        <label
          className={classNames(styles.label, {
            [styles.labelDisabled]: disabled === true,
          })}
          onClick={captureFocus}
        >
          {label}
        </label>
      )}
      <div
        className={classNames(styles.inputWrapper, className, {
          [styles.focused]: focused,
          [styles.errored]: errored,
          [styles.disabled]: disabled === true,
        })}
      >
        {Icon && (
          <div className={styles.iconContainer}>
            <Icon onClick={captureFocus} size={18} />
          </div>
        )}
        <input
          autoFocus={autoFocus}
          className={classNames(styles.input, {
            [styles.withIcon]: !!Icon,
            [styles.withSpinner]: !!loading,
          })}
          disabled={disabled === true}
          onBlur={updateFocusState}
          onChange={onChange}
          onFocus={updateFocusState}
          onKeyDown={e => onKeyDown && onKeyDown(e)}
          onKeyUp={e => onKeyUp && onKeyUp(e)}
          placeholder={placeholder}
          ref={node => {
            inputRef.current = node

            if (innerRef) {
              innerRef.current = node
            }
          }}
          type={type}
          value={value}
        />
        {/* eslint-disable-next-line jsx-a11y/no-static-element-interactions */}
        <div
          className={classNames(styles.loadingWrapper, {
            [styles.loadingVisible]: loading,
          })}
          onClick={captureFocus}
        >
          <Loading size={15} tiny={true} type="spinner" variation={variation} />
        </div>
      </div>
    </div>
  )
}

export default Input
