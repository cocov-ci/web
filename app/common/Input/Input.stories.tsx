import type { Meta, StoryObj } from '@storybook/react'
import { LucideSearch } from 'lucide-react'

import Input from './index'

const meta: Meta<typeof Input> = {
  title: 'Common/Input',
  component: Input,
  tags: ['autodocs'],
  args: {
    width: '200px',
  },
}

export default meta

type Story = StoryObj<typeof Input>

export const Default: Story = {
  args: {
    type: 'text',
  },
}

export const WithPlaceholder: Story = {
  args: {
    type: 'text',
    placeholder: 'Placeholder',
  },
}

export const WithLabel: Story = {
  args: {
    label: 'Email:',
    type: 'text',
    placeholder: 'paul.appleseed@example.com',
  },
}

export const WithLoading: Story = {
  args: {
    label: 'Name:',
    type: 'text',
    value: 'The name',
    loading: true,
  },
}

export const WithIcon: Story = {
  args: {
    type: 'text',
    placeholder: 'Type to Search...',
    loading: true,
    icon: LucideSearch,
  },
}

export const Errored: Story = {
  args: {
    type: 'text',
    errored: true,
  },
}

export const Disabled: Story = {
  args: {
    label: 'Test:',
    type: 'text',
    disabled: true,
  },
}
