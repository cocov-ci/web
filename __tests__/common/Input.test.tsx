import { fireEvent, render, screen } from '@testing-library/react'
import { Search } from 'lucide-react'

import Input from 'app/common/Input'

import '@testing-library/jest-dom'

describe('common/Input', () => {
  it('renders Input', () => {
    const { container } = render(<Input type="text" />)

    expect(container.querySelector('input')).toBeVisible()
  })

  it('renders Input with "width" props', () => {
    const { container } = render(<Input type="text" width="200px" />)

    expect(container.querySelector('div')).toHaveAttribute(
      'style',
      'width: 200px;',
    )
  })

  it('renders Input with "icon" props', () => {
    const { container } = render(<Input icon={Search} type="text" />)

    expect(container.querySelector('svg')).toBeVisible()
  })

  it('renders Input with "value" props', () => {
    const { container } = render(
      <Input onChange={() => null} type="text" value="value here..." />,
    )

    expect(container.querySelector('input')).toHaveValue('value here...')
  })

  it('renders Input with "disabled" props', () => {
    const { container } = render(<Input disabled type="text" />)

    expect(container.querySelector('input')).toHaveAttribute('disabled')
  })

  it('renders Input with "placeholder" props', () => {
    render(<Input placeholder="placeholder here..." type="text" />)

    expect(screen.getByPlaceholderText('placeholder here...')).toBeVisible()
  })

  it('renders Input with "errored" props', () => {
    const { container } = render(<Input errored type="text" />)

    expect(container.querySelector('.errored')).toBeVisible()
  })

  it('renders Input with "label" props', () => {
    const { container } = render(<Input label="label here..." type="text" />)

    expect(container.querySelector('label')).toHaveTextContent('label here...')
  })

  it('renders Input with "loading" props', () => {
    const { container } = render(<Input loading type="text" />)

    expect(container.querySelector('.loadingVisible img')).toBeVisible()
  })

  it('renders Input with "variation" props', () => {
    const { container } = render(<Input type="text" variation="dark" />)

    expect(container.querySelector('div.dark')).toBeVisible()
  })

  it('renders Input with "onChange" props', () => {
    const handleOnChange = jest.fn()
    render(<Input onChange={handleOnChange} type="text" />)
    const input = screen.getByRole('textbox') as HTMLInputElement

    fireEvent.change(input, { target: { value: 'text here...' } })

    expect(handleOnChange).toHaveBeenCalledTimes(1)
    expect(input.value).toBe('text here...')
  })

  it('renders Input with "autoFocus" props', () => {
    render(<Input autoFocus type="text" />)

    const input = screen.getByRole('textbox') as HTMLInputElement
    expect(input).toHaveFocus()
  })

  it('tests onBlur event', () => {
    render(<Input type="text" />)

    const input = screen.getByRole('textbox') as HTMLInputElement
    fireEvent.blur(input)
    expect(input).not.toHaveFocus()
  })

  it('tests capture input focus', () => {
    render(<Input label="label here..." type="text" />)

    const label = screen.getByText('label here...')
    const input = screen.getByRole('textbox') as HTMLInputElement

    fireEvent.click(label)
    expect(input).toHaveFocus()
  })

  it('renders Input with "onKeyDown" props', () => {
    const handleKeyDown = jest.fn()
    render(<Input onKeyDown={handleKeyDown} type="text" />)
    const input = screen.getByRole('textbox') as HTMLInputElement

    fireEvent.keyDown(input, { key: 'A', code: 'KeyA' })

    expect(handleKeyDown).toHaveBeenCalledTimes(1)
  })

  it('renders Input with "onKeyUp" props', () => {
    const handleKeyUp = jest.fn()
    render(<Input onKeyUp={handleKeyUp} type="text" />)
    const input = screen.getByRole('textbox') as HTMLInputElement

    fireEvent.keyUp(input, { key: 'A', code: 'KeyA' })

    expect(handleKeyUp).toHaveBeenCalledTimes(1)
  })

  it('renders Input snapshots', () => {
    const { container } = render(<Input type="text" />)
    expect(container).toMatchSnapshot()
  })
})
