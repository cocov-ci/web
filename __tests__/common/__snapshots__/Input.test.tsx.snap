// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`common/Input renders Input snapshots 1`] = `
<div>
  <div
    class="base light"
    style="width: 100%;"
  >
    <div
      class="inputWrapper"
    >
      <input
        class="input"
        type="text"
        value=""
      />
      <div
        class="loadingWrapper"
      >
        <div
          class="loading center tiny"
        >
          <img
            alt="A spinner indicating an ongoing operation"
            class="spinner"
            data-nimg="1"
            decoding="async"
            height="15"
            loading="lazy"
            src="/icons/spinner.svg"
            srcset="/icons/spinner.svg 1x, /icons/spinner.svg 2x"
            style="color: transparent;"
            width="15"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
